In [2]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.preprocessing import LabelEncoder

In [ ]:
#################### SKIP CELLS to CELL labeled:  "# START - Load dataset" ############################

In [7]:
# *** SKIP CELL ***

# Load dataset 
data = pd.read_csv('static/heart.csv')

data.head(10)

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  MaxHR  \
0   40   M           ATA        140          289          0     Normal    172   
1   49   F           NAP        160          180          0     Normal    156   
2   37   M           ATA        130          283          0         ST     98   
3   48   F           ASY        138          214          0     Normal    108   
4   54   M           NAP        150          195          0     Normal    122   
5   39   M           NAP        120          339          0     Normal    170   
6   45   F           ATA        130          237          0     Normal    170   
7   54   M           ATA        110          208          0     Normal    142   
8   37   M           ASY        140          207          0     Normal    130   
9   48   F           ATA        120          284          0     Normal    120   

  ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0              N      0.0       Up             0  
1              N      1.0     Flat             1  
2              N      0.0       Up             0  
3              Y      1.5     Flat             1  
4              N      0.0       Up             0  
5              N      0.0       Up             0  
6              N      0.0       Up             0  
7              N      0.0       Up             0  
8              Y      1.5     Flat             1  
9              N      0.0       Up             0

In [8]:
# *** SKIP CELL ***

print(data.columns)

Index(['Age', 'Sex', 'ChestPainType', 'RestingBP', 'Cholesterol', 'FastingBS',
       'RestingECG', 'MaxHR', 'ExerciseAngina', 'Oldpeak', 'ST_Slope',
       'HeartDisease'],
      dtype='object')


In [27]:
# *** SKIP CELL ***

# List of columns to convert
columns_to_convert = ['Sex', 'ChestPainType', 'RestingECG', 'ExerciseAngina', 'ST_Slope']

# Loop through the columns and create binary columns
for column in columns_to_convert:
    unique_values = data[column].unique()
    for value in unique_values:
        data[f'{column}_{value.lower()}'] = (data[column] == value).astype(int)

# Drop the original columns
data.drop(columns=columns_to_convert, inplace=True)

print(data.head(10))

KeyError: 'Sex'

In [5]:
# *** SKIP CELL ***

# create a new variable to avoid confusion
data2 = data

data2.head(10)

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  MaxHR  \
0   40   M           ATA        140          289          0     Normal    172   
1   49   F           NAP        160          180          0     Normal    156   
2   37   M           ATA        130          283          0         ST     98   
3   48   F           ASY        138          214          0     Normal    108   
4   54   M           NAP        150          195          0     Normal    122   
5   39   M           NAP        120          339          0     Normal    170   
6   45   F           ATA        130          237          0     Normal    170   
7   54   M           ATA        110          208          0     Normal    142   
8   37   M           ASY        140          207          0     Normal    130   
9   48   F           ATA        120          284          0     Normal    120   

  ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0              N      0.0       Up             0  
1              N      1.0     Flat             1  
2              N      0.0       Up             0  
3              Y      1.5     Flat             1  
4              N      0.0       Up             0  
5              N      0.0       Up             0  
6              N      0.0       Up             0  
7              N      0.0       Up             0  
8              Y      1.5     Flat             1  
9              N      0.0       Up             0

In [23]:
# *** SKIP CELL ***

print(data2.columns)

Index(['Age', 'RestingBP', 'Cholesterol', 'FastingBS', 'MaxHR', 'Oldpeak',
       'HeartDisease', 'Sex_m', 'Sex_f', 'ChestPainType_ata',
       'ChestPainType_nap', 'ChestPainType_asy', 'ChestPainType_ta',
       'RestingECG_normal', 'RestingECG_st', 'RestingECG_lvh',
       'ExerciseAngina_n', 'ExerciseAngina_y', 'ST_Slope_up', 'ST_Slope_flat',
       'ST_Slope_down'],
      dtype='object')


In [24]:
# *** SKIP CELL ***

# data2.to_csv('heart_integer_v01.csv', index=False)

In [ ]:
############################################################

In [ ]:
#################### START BELOW ############################

In [3]:
# START - Load dataset 
data2 = pd.read_csv('static_[null]/heart_integer_v01.csv')

data2.head(10)

Age  RestingBP  Cholesterol  FastingBS  MaxHR  Oldpeak  HeartDisease  \
0   40        140          289          0    172      0.0             0   
1   49        160          180          0    156      1.0             1   
2   37        130          283          0     98      0.0             0   
3   48        138          214          0    108      1.5             1   
4   54        150          195          0    122      0.0             0   
5   39        120          339          0    170      0.0             0   
6   45        130          237          0    170      0.0             0   
7   54        110          208          0    142      0.0             0   
8   37        140          207          0    130      1.5             1   
9   48        120          284          0    120      0.0             0   

   Sex_m  Sex_f  ChestPainType_ata  ...  ChestPainType_asy  ChestPainType_ta  \
0      1      0                  1  ...                  0                 0   
1      0      1                  0  ...                  0                 0   
2      1      0                  1  ...                  0                 0   
3      0      1                  0  ...                  1                 0   
4      1      0                  0  ...                  0                 0   
5      1      0                  0  ...                  0                 0   
6      0      1                  1  ...                  0                 0   
7      1      0                  1  ...                  0                 0   
8      1      0                  0  ...                  1                 0   
9      0      1                  1  ...                  0                 0   

   RestingECG_normal  RestingECG_st  RestingECG_lvh  ExerciseAngina_n  \
0                  1              0               0                 1   
1                  1              0               0                 1   
2                  0              1               0                 1   
3                  1              0               0                 0   
4                  1              0               0                 1   
5                  1              0               0                 1   
6                  1              0               0                 1   
7                  1              0               0                 1   
8                  1              0               0                 0   
9                  1              0               0                 1   

   ExerciseAngina_y  ST_Slope_up  ST_Slope_flat  ST_Slope_down  
0                 0            1              0              0  
1                 0            0              1              0  
2                 0            1              0              0  
3                 1            0              1              0  
4                 0            1              0              0  
5                 0            1              0              0  
6                 0            1              0              0  
7                 0            1              0              0  
8                 1            0              1              0  
9                 0            1              0              0  

[10 rows x 21 columns]

In [4]:
print(data2.columns)

Index(['Age', 'RestingBP', 'Cholesterol', 'FastingBS', 'MaxHR', 'Oldpeak',
       'HeartDisease', 'Sex_m', 'Sex_f', 'ChestPainType_ata',
       'ChestPainType_nap', 'ChestPainType_asy', 'ChestPainType_ta',
       'RestingECG_normal', 'RestingECG_st', 'RestingECG_lvh',
       'ExerciseAngina_n', 'ExerciseAngina_y', 'ST_Slope_up', 'ST_Slope_flat',
       'ST_Slope_down'],
      dtype='object')


In [5]:
# Split into features (X) and target variable (y)
X = data2.drop('HeartDisease', axis=1)
y = data2['HeartDisease']

In [6]:
# SECOND SCALER - STANDARD
# Scale the data, using StandardScaler
from sklearn.preprocessing import StandardScaler

# Create a StandardScaler instance
scaler = StandardScaler()

# Fit and transform the scaler on your feature data
X_scaled1 = scaler.fit_transform(X)

In [7]:
import pandas as pd

# Create a DataFrame from the scaled features
X_scaled1_df = pd.DataFrame(X_scaled1, columns=X.columns)

# Display the first few rows of the scaled dataset
X_scaled1_df.head(10)

Age  RestingBP  Cholesterol  FastingBS     MaxHR   Oldpeak     Sex_m  \
0 -1.433140   0.410909     0.825070  -0.551341  1.382928 -0.832432  0.515952   
1 -0.478484   1.491752    -0.171961  -0.551341  0.754157  0.105664 -1.938163   
2 -1.751359  -0.129513     0.770188  -0.551341 -1.525138 -0.832432  0.515952   
3 -0.584556   0.302825     0.139040  -0.551341 -1.132156  0.574711 -1.938163   
4  0.051881   0.951331    -0.034755  -0.551341 -0.581981 -0.832432  0.515952   
5 -1.539213  -0.669935     1.282424  -0.551341  1.304332 -0.832432  0.515952   
6 -0.902775  -0.129513     0.349422  -0.551341  1.304332 -0.832432 -1.938163   
7  0.051881  -1.210356     0.084157  -0.551341  0.203982 -0.832432  0.515952   
8 -1.751359   0.410909     0.075010  -0.551341 -0.267596  0.574711  0.515952   
9 -0.584556  -0.669935     0.779335  -0.551341 -0.660578 -0.832432 -1.938163   

      Sex_f  ChestPainType_ata  ChestPainType_nap  ChestPainType_asy  \
0 -0.515952           2.075177          -0.532838          -1.084138   
1  1.938163          -0.481887           1.876744          -1.084138   
2 -0.515952           2.075177          -0.532838          -1.084138   
3  1.938163          -0.481887          -0.532838           0.922392   
4 -0.515952          -0.481887           1.876744          -1.084138   
5 -0.515952          -0.481887           1.876744          -1.084138   
6  1.938163           2.075177          -0.532838          -1.084138   
7 -0.515952           2.075177          -0.532838          -1.084138   
8 -0.515952          -0.481887          -0.532838           0.922392   
9  1.938163           2.075177          -0.532838          -1.084138   

   ChestPainType_ta  RestingECG_normal  RestingECG_st  RestingECG_lvh  \
0         -0.229679           0.814275      -0.490449       -0.507478   
1         -0.229679           0.814275      -0.490449       -0.507478   
2         -0.229679          -1.228087       2.038947       -0.507478   
3         -0.229679           0.814275      -0.490449       -0.507478   
4         -0.229679           0.814275      -0.490449       -0.507478   
5         -0.229679           0.814275      -0.490449       -0.507478   
6         -0.229679           0.814275      -0.490449       -0.507478   
7         -0.229679           0.814275      -0.490449       -0.507478   
8         -0.229679           0.814275      -0.490449       -0.507478   
9         -0.229679           0.814275      -0.490449       -0.507478   

   ExerciseAngina_n  ExerciseAngina_y  ST_Slope_up  ST_Slope_flat  \
0          0.823556         -0.823556     1.150674      -1.002181   
1          0.823556         -0.823556    -0.869056       0.997824   
2          0.823556         -0.823556     1.150674      -1.002181   
3         -1.214246          1.214246    -0.869056       0.997824   
4          0.823556         -0.823556     1.150674      -1.002181   
5          0.823556         -0.823556     1.150674      -1.002181   
6          0.823556         -0.823556     1.150674      -1.002181   
7          0.823556         -0.823556     1.150674      -1.002181   
8         -1.214246          1.214246    -0.869056       0.997824   
9          0.823556         -0.823556     1.150674      -1.002181   

   ST_Slope_down  
0      -0.271448  
1      -0.271448  
2      -0.271448  
3      -0.271448  
4      -0.271448  
5      -0.271448  
6      -0.271448  
7      -0.271448  
8      -0.271448  
9      -0.271448

In [33]:
# StandardScaler to CSV
# X_scaled1_df.to_csv('heart_scale_StandardScaler_v01.csv', index=False)

In [8]:
# SECOND SCALER - MINMAX
# Scale the data, using MinMaxScaler
from sklearn.preprocessing import MinMaxScaler

# Create a MinMaxScaler instance
scaler = MinMaxScaler()

# Fit and transform the scaler on your feature data
X_scaled2 = scaler.fit_transform(X)

In [9]:
import pandas as pd

# Create a DataFrame from the scaled features
X_scaled2_df = pd.DataFrame(X_scaled2, columns=X.columns)

# Display the first few rows of the scaled dataset
X_scaled2_df.head(10)

Age  RestingBP  Cholesterol  FastingBS     MaxHR   Oldpeak  Sex_m  \
0  0.244898       0.70     0.479270        0.0  0.788732  0.295455    1.0   
1  0.428571       0.80     0.298507        0.0  0.676056  0.409091    0.0   
2  0.183673       0.65     0.469320        0.0  0.267606  0.295455    1.0   
3  0.408163       0.69     0.354892        0.0  0.338028  0.465909    0.0   
4  0.530612       0.75     0.323383        0.0  0.436620  0.295455    1.0   
5  0.224490       0.60     0.562189        0.0  0.774648  0.295455    1.0   
6  0.346939       0.65     0.393035        0.0  0.774648  0.295455    0.0   
7  0.530612       0.55     0.344942        0.0  0.577465  0.295455    1.0   
8  0.183673       0.70     0.343284        0.0  0.492958  0.465909    1.0   
9  0.408163       0.60     0.470978        0.0  0.422535  0.295455    0.0   

   Sex_f  ChestPainType_ata  ChestPainType_nap  ChestPainType_asy  \
0    0.0                1.0                0.0                0.0   
1    1.0                0.0                1.0                0.0   
2    0.0                1.0                0.0                0.0   
3    1.0                0.0                0.0                1.0   
4    0.0                0.0                1.0                0.0   
5    0.0                0.0                1.0                0.0   
6    1.0                1.0                0.0                0.0   
7    0.0                1.0                0.0                0.0   
8    0.0                0.0                0.0                1.0   
9    1.0                1.0                0.0                0.0   

   ChestPainType_ta  RestingECG_normal  RestingECG_st  RestingECG_lvh  \
0               0.0                1.0            0.0             0.0   
1               0.0                1.0            0.0             0.0   
2               0.0                0.0            1.0             0.0   
3               0.0                1.0            0.0             0.0   
4               0.0                1.0            0.0             0.0   
5               0.0                1.0            0.0             0.0   
6               0.0                1.0            0.0             0.0   
7               0.0                1.0            0.0             0.0   
8               0.0                1.0            0.0             0.0   
9               0.0                1.0            0.0             0.0   

   ExerciseAngina_n  ExerciseAngina_y  ST_Slope_up  ST_Slope_flat  \
0               1.0               0.0          1.0            0.0   
1               1.0               0.0          0.0            1.0   
2               1.0               0.0          1.0            0.0   
3               0.0               1.0          0.0            1.0   
4               1.0               0.0          1.0            0.0   
5               1.0               0.0          1.0            0.0   
6               1.0               0.0          1.0            0.0   
7               1.0               0.0          1.0            0.0   
8               0.0               1.0          0.0            1.0   
9               1.0               0.0          1.0            0.0   

   ST_Slope_down  
0            0.0  
1            0.0  
2            0.0  
3            0.0  
4            0.0  
5            0.0  
6            0.0  
7            0.0  
8            0.0  
9            0.0

In [34]:
# MinMaxScaler to CSV
# X_scaled2_df.to_csv('heart_scale_MinMaxScaler_v01.csv', index=False)

In [10]:
# Split the data into training and testing sets
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X_scaled2, y, test_size=0.2, random_state=42)

In [11]:
from sklearn.tree import DecisionTreeClassifier

# Create a Decision Tree Classifier instance
dt_classifier = DecisionTreeClassifier(random_state=42)

# Train the classifier on the training data
dt_classifier.fit(X_train, y_train)

DecisionTreeClassifier(random_state=42)

In [12]:
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# Make predictions on the testing data
y_pred = dt_classifier.predict(X_test)

print("----------------------------------------------------------")

# Calculate accuracy
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

print("----------------------------------------------------------")

# Other classification metrics
print(classification_report(y_test, y_pred))

print("----------------------------------------------------------")

# Confusion matrix
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred))

print("----------------------------------------------------------")

----------------------------------------------------------
Accuracy: 0.7880434782608695
----------------------------------------------------------
              precision    recall  f1-score   support

           0       0.71      0.84      0.77        77
           1       0.87      0.75      0.80       107

    accuracy                           0.79       184
   macro avg       0.79      0.80      0.79       184
weighted avg       0.80      0.79      0.79       184

----------------------------------------------------------
Confusion Matrix:
 [[65 12]
 [27 80]]
----------------------------------------------------------


Summary analysis of the Decision Tree Classifier model's performance:

**Accuracy:** The model achieved an accuracy of approximately 77.17%. This metric measures the proportion of correctly predicted instances out of all instances in the test set.

**Precision:** The model's precision is about 75% for class 0 (no heart disease) and 78.43% for class 1 (heart disease). Precision measures the ability of the model to make correct positive predictions. For class 0, it means that 75% of the predicted instances as having no heart disease were correct. For class 1, it means that 78.43% of the predicted instances as having heart disease were correct.

**Recall:** The model's recall is approximately 74.77% for class 0 and 78.50% for class 1. Recall measures the ability of the model to identify all relevant instances. For class 0, it means that 74.77% of actual instances with no heart disease were correctly identified. For class 1, it means that 78.50% of actual instances with heart disease were correctly identified.

**F1-Score:** The F1-score is around 75.38% for class 0 and 78.47% for class 1. The F1-score is the harmonic mean of precision and recall and provides a balanced measure of a model's performance.

**Confusion Matrix:** The confusion matrix provides a more detailed breakdown of the model's predictions. It shows that the model correctly predicted 58 instances with no heart disease and 90 instances with heart disease. However, it also made 20 false-positive predictions (predicted as having heart disease but actually didn't have) and 16 false-negative predictions (predicted as not having heart disease but actually did have).

In summary, the Decision Tree Classifier model exhibits decent performance with an accuracy of around 77%. However, there is room for improvement, particularly in reducing the number of false-positive and false-negative predictions. Can experiment with different hyperparameters or try other classification algorithms to enhance the model's performance further.
